cmd/check/main.go

```diff
@@ -64,19 +65,31 @@ func main() {
 			return
 		}
 
+		origin = new(name.Registry)
+		*origin = repo.Registry
 		repo.Registry = registry
 	}
 
 	ref := repo.Tag(req.Source.Tag())
 
 	var response CheckResponse
-	err = resource.RetryOnRateLimit(func() error {
+	op := func() error {
 		var err error
 		response, err = check(req, ref)
 		return err
-	})
+	}
+	err = resource.RetryOnRateLimit(op)
+	if err != nil {
+		logrus.Warnf("checking mirror %s failed: %s", ref.RegistryStr(), err)
```

Comment:
The business logic makes sense, but, this would show an incorrect log messages if a `mirror` isn't configured and the `origin` fails.
Pulling out the `op` function looks good, I am wondering if the function would be easier to read if the `mirror` logic as under a single `if` block ?

Reply:
> this would show an incorrect log messages if a mirror isn't configured and the origin fails.

Good call. Definitely needs an update.

> I am wondering if the function would be easier to read if the mirror logic [was] under a single if block ?

The part that I'm struggling with here is how you signal the need for a retry against the origin. In order to move the `RetryOnRateLimit` call up in to the `if` block, we need a way to determine if another call to `RetryOnRateLimit` is needed (and skipping the follow on try if the first was successful).

I'll play around with it a bit, but if you have something in mind and can sketch out an example, that would help me follow what you're thinking.